### ISSUE TITLE
fix(lora): make adapter weight fusion Conv1D-layout aware

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#925]
## Problem / Background

`fuse_lora_weights_into` fuses a LoRA delta into a base tensor with a bare element-wise add and no shape check:

```rust
// src/lora/loader.rs:149
let fused = mlxcel_core::add(base_weight, &delta);
```

`compute_lora_delta` always produces a delta in `mlxcel_core` `Linear` layout, `[out_features, in_features]`. Fusion has no notion of how the base tensor is actually stored, and it never validates that the two operands agree. That held up while every supported family stored 2-D projections as `[out, in]`.

GPT-2 (#833, PR #924) breaks that assumption. It is the first family in this tree whose raw HuggingFace exports store `c_attn`, `c_proj`, and `c_fc` in the HF `Conv1D` layout `[in, out]`, transposed relative to what `Linear` expects. `Gpt2Layout::detect` and `Gpt2Layout::conv1d_linear` in `src/models/gpt2.rs` handle this with a load-time transpose, and they do it correctly.

The reachability comes from ordering. In `load_model_with_adapter` (`src/loading/mod.rs`), the raw safetensors map is loaded, `lora::apply_lora_adapters` runs on it, and only afterwards does `load_model_from_weights` construct the model, which is where `Gpt2Layout` performs the transpose. So at the moment fusion runs against a raw HF GPT-2 checkpoint, the base weight is still in `[in, out]` Conv1D layout while the delta is in `[out, in]`. GPT-2 is adapter-enabled: `src/model_metadata.rs` registers `Gpt2` with `adapter: None`, meaning there is no adapter-unsupported message, so nothing gates the path. The pipeline-parallel entry point `apply_stage_lora_adapter` funnels through the same `fuse_lora_weights_into` primitive and has the same exposure.

### Concrete outcomes on a raw HF GPT-2 checkpoint

For `n_embd = H` (H = 768 on GPT-2 small):

| Base weight | Raw HF (Conv1D) shape | Delta shape | Result of `add` |
|---|---|---|---|
| `h.N.attn.c_attn.weight` | `[H, 3H]` = `[768, 2304]` | `[3H, H]` = `[2304, 768]` | broadcast failure |
| `h.N.mlp.c_fc.weight` | `[H, 4H]` = `[768, 3072]` | `[4H, H]` = `[3072, 768]` | broadcast failure |
| `h.N.mlp.c_proj.weight` | `[4H, H]` = `[3072, 768]` | `[H, 4H]` = `[768, 3072]` | broadcast failure |
| `h.N.attn.c_proj.weight` | `[H, H]` = `[768, 768]` | `[H, H]` = `[768, 768]` | **succeeds, silently wrong** |

The square attention `c_proj` is the case that motivates this issue. Writing the true Linear weight as `W` (`[out, in]`), the checkpoint holds `Wᵀ`, so fusion yields `Wᵀ + D`. `conv1d_linear` then transposes at construction time, and the model ends up running `W + Dᵀ` instead of `W + D`. `Dᵀ` is a different rank-r update of the same shape, so there is no error, no warning, and no shape anomaly. The adapter appears to load and the model produces plausible but incorrect output.

The three non-square projections fail hard, but not cleanly. `add` is declared in the cxx bridge as `fn add(a: &MlxArray, b: &MlxArray) -> UniquePtr<MlxArray>` with no `Result`, and the C++ side (`src/lib/mlxcel-core/cpp/mlx_cxx_bridge.cpp:430`) calls `mlx::core::add` with no try/catch. The MLX broadcast error is a C++ exception escaping a cxx shim that is not declared fallible, so it terminates the process rather than surfacing as an `anyhow::Error` the caller can report.

### Why this is narrow in practice today

Reaching the bug requires an adapter whose keys match the bare GPT-2 projection names against a raw (untransposed) checkpoint. The two adapter formats in circulation do not:

- **mlx-lm adapters** carry the `model.` prefix, matching the MLX-converted GPT-2 layout. A `model.`-prefixed GPT-2 checkpoint in circulation is an MLX conversion and is already stored `[out, in]`, so those adapters resolve only against weights that need no transpose. Note this is a property of how the published conversions are packed, not something the loader enforces: `Gpt2Layout::detect` decides the layout purely from the shape of `h.0.attn.c_attn.weight` and never consults the prefix, so a `model.`-prefixed checkpoint that was still in Conv1D layout would be detected as Conv1D and would be exposed exactly like the bare form.
- **HuggingFace PEFT adapters** name their tensors `lora_A.weight` / `lora_B.weight`. The key scan in `fuse_lora_weights_into` matches on `.lora_a` / `.lora_b` suffixes, so PEFT keys never match at all and fusion is a no-op for them.

So the trigger is an adapter using mlxcel's own `.lora_a` / `.lora_b` naming on bare `h.N.attn.c_proj`-style keys. That is a real and producible artifact (it is exactly what `apply_stage_lora_adapter` and the local fine-tuning conventions describe), but it is not what today's published adapters look like. The defect is latent rather than actively breaking users, which is why the priority is low. It is worth fixing because the failure mode on the square projection is silent numerical corruption, and it becomes progressively more likely as GPT-2 adapter tooling appears.

### Re-verified on current `main` (`a3c823611`)

Every claim above was re-checked after #924, #926, #928, #930 and #946 merged. Nothing moved:

- `src/lora/loader.rs:149` is still the unguarded `mlxcel_core::add(base_weight, &delta)`. `compute_lora_delta` still normalizes both adapter conventions to `[out, in]` (transposing the `a @ b` product on the mlx-lm branch, computing `b @ a` on the PEFT branch), and `find_base_weight_name` still falls through to `{name}.weight` when no candidate matches.
- `apply_stage_lora_adapter` (`src/lora/loader.rs:298`) still calls the same primitive, at `src/lora/loader.rs:334`, so the pipeline-parallel path has identical exposure and no guard of its own.
- `add` is still declared without `Result` at `src/lib/mlxcel-core/src/lib.rs:219`, and the shim at `src/lib/mlxcel-core/cpp/mlx_cxx_bridge.cpp:430` is still a bare `mlx::core::add` with no try/catch. Only 7 declarations in the whole bridge return `Result`, and `add` is not one of them.
- `src/model_metadata.rs:204` still registers `Gpt2` with `adapter: None`, so the `adapter_unsupported_message` gate in `load_model_from_weights` (`src/loading/mod.rs:656`) has nothing to reject on.
- GPT-2 is still the only Conv1D-layout family in the tree. `src/models/gpt_bigcode.rs:40-48` documents that GPT-BigCode builds its projections with `nn.Linear`, and its `load_linear` rejects a `[in, out]` weight by name with a message that points at the GPT-2 orientation. GPT-NeoX, Helium and Bailing MoE all store `[out, in]`. The guard is still worth having for the next family that lands transposed, but no family other than GPT-2 is exposed today.

## Proposed Solution

Make weight fusion layout-aware, and make it fail loudly rather than guess.

1. **Shape-check before `add`.** In `fuse_lora_weights_into`, compare `array_shape(base_weight)` against `array_shape(&delta)` before the add. On a mismatch, return an `anyhow::Error` naming the base weight key and both shapes. This alone converts three of the four GPT-2 cases from a process abort into a diagnosable error.
2. **Reject, do not auto-transpose.** If the base shape is the reverse of the delta shape, that is ambiguous evidence, not a repair instruction: for a square weight the two are indistinguishable, so silently transposing would fix `c_fc` while leaving `c_proj` just as wrong. Report the layout mismatch and refuse the fusion.
3. **Give fusion the layout it needs.** Options, in rough order of preference:
   - Normalize the checkpoint layout before fusion, so `load_model_with_adapter` hands `fuse_lora_weights_into` a weight map that is already in `[out, in]`. This is the cleanest fix because it removes the ordering hazard entirely rather than patching around it, and it keeps the fusion primitive family-agnostic.
   - Failing that, thread a per-key layout hint into fusion so it can transpose the delta (not the base) for known Conv1D keys.
4. **Extend the guard to the pipeline-parallel path.** `apply_stage_lora_adapter` shares the primitive, so it inherits the fix, but the stage-local path needs its own coverage since it never goes through `load_model_with_adapter`.

Whatever shape the fix takes, a square Conv1D-layout weight must not be able to accept a transposed delta without an error.

## Acceptance Criteria

- [x] `fuse_lora_weights_into` validates operand shapes before the element-wise add and returns a descriptive `anyhow::Error` (naming the base weight key, the base shape, and the delta shape) instead of letting MLX broadcast rules decide.
- [x] A base/delta shape mismatch never reaches `mlxcel_core::add`, so no LoRA fusion path can abort the process through an uncaught C++ exception.
- [x] A square base weight stored in Conv1D `[in, out]` layout cannot be fused with an `[out, in]` delta without an error. Silent transposed accumulation is eliminated, not merely made unlikely.
- [x] Fusion never silently transposes an operand to make shapes agree.
- [x] A real test fuses an adapter using `.lora_a` / `.lora_b` keys against an actual Conv1D-layout GPT-2 checkpoint (raw HuggingFace `gpt2`, bare `h.N.*` keys, no `model.` prefix) and asserts the outcome. Synthetic shape-only unit tests are acceptable as supplementary coverage but are not sufficient on their own, since the whole defect is that the shapes look fine.
- [x] The test explicitly covers the square `h.N.attn.c_proj.weight` case, which is the one that produces no error today.
- [x] The non-square projections (`c_attn`, `c_fc`, and the MLP `c_proj`) produce the new descriptive error rather than a crash.
- [ ] Whichever fix direction is chosen, the end state is verified end to end: loading a GPT-2 checkpoint with an adapter through `load_model_with_adapter` either produces numerically correct fused weights or a clear error, and generation runs against a real checkpoint to confirm it.
- [x] `apply_stage_lora_adapter` (pipeline-parallel stage-local fusion) is covered by the same guarantee and has its own test.
- [x] Existing LoRA fusion behavior for `[out, in]` families is unchanged, verified against the current adapter tests in `src/distributed/pipeline/partial_loading_adapter_tests.rs`.

## Technical Considerations

- Relevant code: `src/lora/loader.rs` (`fuse_lora_weights_into`, `compute_lora_delta`, `find_base_weight_name`, `apply_stage_lora_adapter`), `src/loading/mod.rs` (`load_model_with_adapter`), `src/models/gpt2.rs` (`Gpt2Layout`), `src/lib/mlxcel-core/cpp/mlx_cxx_bridge.cpp` (`add`).
- `find_base_weight_name` resolves `h.0.attn.c_proj` to `h.0.attn.c_proj.weight` on a raw checkpoint, and falls through to that same candidate even when no key matches. Any shape guard has to sit after resolution, on the tensor that was actually found.
- The two "base weight not found" and "incomplete LoRA pair" paths in `fuse_lora_weights_into` report through `tracing::warn!`, which is a no-op in the `mlxcel` CLI binary because only `src/server/startup.rs` installs a subscriber. Any new diagnostic on the CLI-reachable path has to be a returned error or an `eprintln!` to be visible at all.
- `compute_lora_delta` already infers the adapter convention (mlx-lm versus PEFT) from shapes. Adding a base-layout inference on top of that stacks one heuristic on another, which argues for normalizing the checkpoint layout ahead of fusion instead.
- Quantized GPT-2 checkpoints short-circuit the Conv1D probe in `Gpt2Layout::detect` (a packed weight shape carries no layout signal, and a quantized export can only come from an already-sanitized MLX conversion). Fusion against quantized base weights is a separate concern and is out of scope here.
- This is not GPT-2-specific in principle. Any future family that lands with a transposed on-disk projection layout hits the same primitive, so the guard is worth having independent of how the GPT-2 case is resolved.

<details>
<summary><b>Implementation context (machine facts, validation method, and a recurring defect class)</b></summary>

**Most relevant to this issue:** that MLX `slice` clamps an out-of-range stop instead of throwing (the `c_attn` three-way split rides on that behavior), that a C++ throw crossing the cxx bridge is an uncatchable `std::terminate` rather than a recoverable error (which is why the three non-square projections abort the process instead of returning an error), and the reminder to update the `// Used by:` comment when touching a shared helper, since `fuse_lora_weights_into` carries one.

Collected while porting five text-model families in #924, #926, #928, #930, and #946. This
is background for whoever picks this up, not additional scope.

### Build and test on the current development machine (Apple M1 Ultra, macOS)

- Every cargo invocation needs `DEVELOPER_DIR=/Applications/Xcode-26.6.0.app/Contents/Developer`, or cmake fails with `xcrun: error: unable to find utility "metal"`. `xcode-select` still points at CommandLineTools.
- Only the RELEASE profile is warm. A debug `cargo check` triggers a COLD MLX C++ build that takes many minutes. Use `DEVELOPER_DIR=... cargo check --release --lib --tests --features metal,accelerate`.
- Narrow test selectors only: `DEVELOPER_DIR=... cargo test --release --lib --features metal,accelerate models::<module>`. A bare `cargo test`, `cargo test --lib`, or `cargo clippy --workspace --all-targets` runs long enough to stall an agent with a stream-idle watchdog, and a cold `cargo build --release` can exceed it on the MLX C++ link step alone.
- `cargo fmt --all -- --check`, never a bare-path fmt check, which produces spurious edition-2024 diffs.
- CI runs no `cargo test` at all (only `changes`, `cargo-deny`, `cargo-fmt`, and the cross-repo-ref guard). Local runs plus a real checkpoint are the only gate. Two deterministically failing tests were found sitting on `main` during this series for exactly that reason: see #939, and the `FAMILY_ORDER` entries fixed in #946.
- `tracing::warn!` is a NO-OP in the `mlxcel` CLI binary; only `src/server/startup.rs` installs a subscriber. Use `eprintln!` for any CLI-facing diagnostic.
- A test that trips an MLX C++ throw aborts the whole test binary with SIGABRT rather than failing cleanly. That is expected, and the abort is stronger evidence than a clean assertion failure. One SIGABRT on the first run after a fresh link that does not reproduce on reruns is a cold Metal-initialization race, observed independently by two reviewers.

### Validating model behavior: a token-exact reference oracle

Shape tests do not catch the failures that matter here. A wrong prefill/decode offset, a flat instead of interleaved QKV split, the wrong RoPE convention, or a routing bias applied to the wrong copy all produce correctly-shaped tensors and fluent, plausible output. Only the token id sequence separates them.

What worked: create a scratch venv, `pip install mlx-lm`, and drive `mlx_lm.generate.stream_generate` with `make_sampler(temp=0.0)`, printing both text and token ids. Then compare against `mlxcel generate`. **Pass `--no-chat-template` to mlxcel even when the checkpoint ships a chat template**, so both sides see the same raw prompt, and run the templated path separately as a usability check.

Caveat: the oracle is blind wherever mlxcel and mlx-lm make the same choice. GELU was one such case, closed by monkey-patching the reference to the other variant and regenerating. Note `mlxcel_core::utils::gelu_approx` is erf-based despite its name, while MLX Python's `nn.gelu_approx` really is the tanh form.

### A recurring defect class worth checking for

A value from `config.json` passes every Rust-side check, violates an undocumented precondition of an MLX C++ entry point, and kills the process at the FIRST FORWARD PASS rather than failing at load. Most cxx bridge functions are declared returning `UniquePtr<MlxArray>` rather than `Result`, so the C++ throw is an uncatchable `std::terminate`, not something `catch_unwind` contains. The model loads cleanly and the server dies on its first request. This class produced one CRITICAL and five HIGH findings across the five ports.

Specifics established by reading the pinned MLX checkout rather than assuming:

- Gathers do NOT range-check positive indices. `take` wraps negative indices but an out-of-range positive index silently returns values belonging to no row, and that reaches the logits with no fault. Bound every gather by the real tensor shape, never by a config field.
- `slice` CLAMPS an out-of-range stop instead of throwing, so a too-wide split silently loses trailing channels.
- `fast_rope` requires `dims` even, positive, and no larger than the last axis.
- `rms_norm` and `layer_norm` never inspect `eps`. A NaN or negative eps yields NaN hidden states with no error at all, which is harder to diagnose than a crash.
- `matmul` throws on an inner-dimension mismatch, which is why an unchecked INPUT axis is fatal even where the row axis bounds only an argmax.
- `quantized_matmul` divides by `bits`, so `"bits": 0` is a division by zero. On AArch64 the divide returns 0 and `std::invalid_argument` fires; on x86-64 the hardware raises `SIGFPE` and kills the process before any exception exists.
- Reconstruct a quantized input width the way MLX does, as `scales.shape(-1) * group_size`, and check `.biases` shape equals `.scales` shape. A self-consistency check cannot catch a checkpoint honestly packed for the wrong `hidden_size`, because such a checkpoint IS self-consistent.
- Put zero checks BEFORE divisibility checks, because `0.is_multiple_of(0)` is true.
- Avoid unbounded probe loops over `0..n_layer` in the load path: a huge `n_layer` hangs with a flat allocation footprint, so no OOM kill rescues it.

`src/models/gpt_neox.rs` and `src/models/helium.rs` carry the current `ModelArgs::validate` / `validate_weights` shape to copy from.

### Conventions that bite

- The `// Used by:` comments above shared helpers are the designated discovery mechanism for "what breaks if I change this", and they can themselves be stale. One was found listing six callers where there were eight. Verify by grep rather than trusting the comment, and update it when you touch the helper.
- `mlxcel arch` is the architecture registry. `mlxcel list` lists downloaded models and will not tell you whether a family is supported.

</details>
